src/test/java/org/apache/mesos/offer/constrain/AndRuleTest.java

```diff
@@ -0,0 +1,207 @@
+        b.getSetBuilder().addItem("foo").addItem("bar").addItem("baz");
+        RESOURCE_3 = b.build();
+
+        b = Resource.newBuilder().setName("4").setType(Value.Type.RANGES);
+        b.getRangesBuilder().addRangeBuilder().setBegin(7).setEnd(8);
+        b.getRangesBuilder().addRangeBuilder().setBegin(10).setEnd(12);
+        RESOURCE_4 = b.build();
+
+        RESOURCES = Arrays.asList(RESOURCE_1, RESOURCE_2, RESOURCE_3, RESOURCE_4);
+    }
+
+    private static final PlacementRule ALL = new PlacementRule() {
+        @Override
+        public Offer filter(Offer offer) {
+            return offer;
+        }
+    };
+    private static final PlacementRule NONE = new PlacementRule() {
+        @Override
+        public Offer filter(Offer offer) {
+            return offer.toBuilder().clearResources().build();
+        }
+    };
+    private static final PlacementRule REMOVE_FIRST = new PlacementRule() {
+        @Override
+        public Offer filter(Offer offer) {
+            if (offer.getResourcesCount() == 0) {
+                return offer;
+            }
+            return offer.toBuilder().removeResources(0).build();
+        }
+    };
+    private static final PlacementRule REMOVE_LAST = new PlacementRule() {
+        @Override
+        public Offer filter(Offer offer) {
+            int count = offer.getResourcesCount();
+            if (count == 0) {
+                return offer;
+            }
+            return offer.toBuilder().removeResources(count - 1).build();
+        }
+    };
+
+    @Test
+    public void testEmpty() {
+        Offer o = new AndRule().filter(offerWith(RESOURCES));
+        assertTrue(o.getResourcesList().isEmpty());
+    }
+
+    @Test
+    public void testAllPass() {
+        Offer o = new AndRule(ALL).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCES), o);
+
+        o = new AndRule(ALL, ALL, ALL).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCES), o);
+    }
+
+    @Test
+    public void testAllFail() {
+        Offer o = new AndRule(NONE).filter(offerWith(RESOURCES));
+        assertTrue(o.getResourcesList().isEmpty());
+
+        o = new AndRule(NONE, ALL).filter(offerWith(RESOURCES));
+        assertTrue(o.getResourcesList().isEmpty());
+
+        o = new AndRule(ALL, NONE).filter(offerWith(RESOURCES));
+        assertTrue(o.getResourcesList().isEmpty());
+
+        o = new AndRule(NONE, NONE).filter(offerWith(RESOURCES));
+        assertTrue(o.getResourcesList().isEmpty());
+
+        o = new AndRule(NONE, ALL, NONE).filter(offerWith(RESOURCES));
+        assertTrue(o.getResourcesList().isEmpty());
+
+        o = new AndRule(ALL, NONE, ALL).filter(offerWith(RESOURCES));
+        assertTrue(o.getResourcesList().isEmpty());
+
+        o = new AndRule(NONE, NONE, NONE).filter(offerWith(RESOURCES));
+        assertTrue(o.getResourcesList().isEmpty());
+    }
+
+    @Test
+    public void testFirstRemoved() {
```

Comment:
Not sure if this is the most descriptive name for the test -- you're not really testing the `REMOVE_FIRST` rule so much as you're testing that `AndRule` properly returns the intersection of several non-trivial (i.e. not `NONE` or `ALL`) placement rules.